### PR TITLE
new plugins: Add non-strict affinity groups

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR adds two VM affinity groups

- Non-strict host affinity
- Non-strict host anti-affinity

This PR also adds a new parameter `considerlasthost` to startVirtualMachine API.

- if `considerlasthost` is false, CloudStack will not consider the last host and start the VM by normal process.
- `considerlasthost` is true by default, CloudStack will try to start the VM on the last host where it run if destination host is not specified.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm genreally. Your solution will work quite well , for this feature
As discussed off-line, a numeric implementation of priority could give more possibilities to bigger clouds:
- for affinity you could give the host/cluster that already has the highest number of VMs the highest prio
- for anti affinity you could take into account distance (in latency or by an operator created map
I think these are add on features but we should consider: will it be hard to change to numeric once this is in?

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java [hunk 1]
```diff
@@ -406,7 +410,7 @@ public DeployDestination planDeployment(VirtualMachineProfile vmProfile, Deploym
             planner = getDeploymentPlannerByName(plannerName);
         }
 
-        if (vm.getLastHostId() != null && haVmTag == null) {
+        if (vm.getLastHostId() != null && haVmTag == null && !forgetLastHostBool) {
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java]: exacted to a variable, hope it is good to you

**Comment by weizhouapache** [comment]
@blueorangutan test

**Review by DaanHoogland — COMMENTED** [review]
reviewing the test I see that I don't understand your use of the Priority enum. It seems more like a PriorityAdjustment enum. Is that correct?

**Comment by weizhouapache** [comment]
@blueorangutan package

File: api/src/main/java/com/cloud/deploy/DataCenterDeployment.java [hunk 2]
```diff
@@ -124,4 +127,30 @@ public String toString() {
...
+    @Override
+    public void addHostPriority(Long hostId, HostPriority priority) {
+        Integer currentPriority = hostPriorities.get(hostId);
+        if (currentPriority == null) {
+            currentPriority = DEFAULT_HOST_PRIORITY;
+        }
+        if (HostPriority.HIGH.equals(priority)) {
+            hostPriorities.put(hostId, currentPriority + 1);
+        } else if (HostPriority.LOW.equals(priority)) {
+            hostPriorities.put(hostId, currentPriority - 1);
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply api/src/main/java/com/cloud/deploy/DataCenterDeployment.java]: with the new code, the priorities are number (positive, negative or 0), not HIGH/LOW/NORMAL
examples can be found at http://qa.cloudstack.cloud/docs/WIP-PROOFING/pr/291/adminguide/virtual_machines.html#determine-destination-host-of-vms-with-non-strict-affinity-groups

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Comment by vladimirpetrov** [comment]
@blueorangutan package

**Review by alexandremattioli — CHANGES_REQUESTED** [review]
options: ['host anti-affinity (Strict)', 'host affinity (Strict)', 'host anti-affinity (Non-Strict)', 'host affinity (Non-Strict)']

Missing hyphens in Non-Strict

**Comment by weizhouapache** [comment]
> options: ['host anti-affinity (Strict)', 'host affinity (Strict)', 'host anti-affinity (Non-Strict)', 'host affinity (Non-Strict)']
> 
> Missing hyphens in Non-Strict

thanks @alexandremattioli , updated

@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java [hunk 3]
```diff
@@ -82,6 +82,13 @@ public class StartVMCmd extends BaseAsyncCmd implements UserCmd {
                since = "3.0.1")
     private Long hostId;
 
+    @Parameter(name = ApiConstants.CONSIDER_LAST_HOST,
+            type = CommandType.BOOLEAN,
+            description = "True by default, CloudStack will firstly try to start the VM on the last host where it run on before stopping, if destination host is not specified. " +
+                    "If false, CloudStack will not consider the last host and start the VM by normal process.",
+            since = "4.18.0")
+    private Boolean considerLastHost;
```
Comment (JoaoJandre) [inline api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java:90]: The default value should be false, to maintain current behavior.

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java [hunk 4]
```diff
@@ -406,7 +408,10 @@ public DeployDestination planDeployment(VirtualMachineProfile vmProfile, Deploym
             planner = getDeploymentPlannerByName(plannerName);
         }
 
-        if (vm.getLastHostId() != null && haVmTag == null) {
+        String considerLastHostStr = (String)vmProfile.getParameter(VirtualMachineProfile.Param.ConsiderLastHost);
+        boolean considerLastHost = vm.getLastHostId() != null && haVmTag == null &&
+                (considerLastHostStr == null || Boolean.TRUE.toString().equalsIgnoreCase(considerLastHostStr));
```
Comment (JoaoJandre) [inline server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:413]:
```suggestion
        boolean considerLastHost = vm.getLastHostId() != null && haVmTag == null &&
                Boolean.TRUE.toString().equalsIgnoreCase(considerLastHostStr);
```
I think we should maintain the current default behavior. That is, if this parameter is not informed, the considerLastHost should be false.

**Review by JoaoJandre — COMMENTED** [review]
I think this PR should be separated in two, as there are two different changes being proposed. This way we can keep things more organized and easier to review.

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java (hunk 4 above)
Comment (alexandremattioli) [reply server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:413]: What's the thought behind it? Any downsides on considering last host?

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java (hunk 3 above)
Comment (weizhouapache) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java:90]: @JoaoJandre 
if you have looked into the code, you may find that the last host is considered as the first option when start a vm.
the current behavior is same as considerLastHost=true.

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java (hunk 4 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:413]: @JoaoJandre 
see above. the current behavior is same as considerLastHost=true

**Comment by DaanHoogland** [comment]
@alexandremattioli according to github you requested changes on this PR but I can find your request. Can you point those out, please? (or re-review?)

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java (hunk 3 above)
Comment (JoaoJandre) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java:90]: Ok, I analyzed the code and what you're saying makes sense.

File: server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java (hunk 4 above)
Comment (JoaoJandre) [reply server/src/main/java/com/cloud/deploy/DeploymentPlanningManagerImpl.java:413]: Same as above.

File: api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java (hunk 3 above)
Comment (weizhouapache) [reply api/src/main/java/org/apache/cloudstack/api/command/user/vm/StartVMCmd.java:90]: thanks @JoaoJandre for reviews

**Comment by weizhouapache** [comment]
@blueorangutan package

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by alexandremattioli** [comment]
LGTM

**Review by alexandremattioli — APPROVED** [review]

**Pull request merged**